File: python/service/admin_token.py
from hashlib import md5
from config.env import Env
from library.safety import Safety
from library.redis import Redis
from util.util import Util
from model.sys_menu import SysMenu

# 后台Token
class AdminToken:

  # 验证
  def verify(self, token: str, urlPerm: str):
    # Token
    if token=='' : return 'Token不能为空!'
    tData = Safety.Decode(token)
    if not tData : return 'Token验证失败!'
    # 是否过期
    uid = str(tData['uid'])
    redis = Redis()
    time = redis.Ttl(Env.admin_token_prefix+'_token_'+uid)
    redis.Close()
    if time <1 : return 'Token已过期!'
    # 续期
    if Env.admin_token_auto :
      redis = Redis()
      redis.Expire(Env.admin_token_prefix+'_token_'+uid, Env.admin_token_time)
      redis.Expire(Env.admin_token_prefix+'_perm_'+uid, Env.admin_token_time)
      redis.Close()
    # URL权限
    if urlPerm=='' : return ''
    arr = Util.Explode('/', urlPerm)
    action = arr[-1:][0]
    controller = Util.Implode('/', arr[:-1])
    # 菜单
    menu = SysMenu()
    menu.Columns('id', 'action')
    menu.Where('controller=%s', controller)
    menuData = menu.FindFirst()
    if not menuData : return '菜单验证无效!'
    # 验证-菜单
    id = str(menuData['id'])
    permData = self.perm(token)
    if id not in permData.keys() : return '无权访问菜单!'
    # 验证-动作
    actionVal = permData[id]
    permArr = Util.JsonDecode(menuData['action'])
    permVal = 0
    for val in permArr :
      if action==val['action'] :
        permVal = int(val['perm'])
        break
    if actionVal&permVal==0 : return '无权访问动作!'
    return ''

  # 权限数组
  def perm(self, token: str):
    # Token
    tData = Safety.Decode(token)
    if not tData : return 'Token验证失败!'
    # 权限
    redis = Redis()
    permStr = redis.Get(Env.admin_token_prefix+'_perm_'+tData['uid'])
    redis.Close()
    # 拆分
    permAll = {}
    arr = [] if not permStr else Util.Explode(' ', permStr)
    for val in arr :
      s = Util.Explode(':', val)
      permAll[s[0]] = int(s[1])
    return permAll

  # 生成
  def create(self, data: dict):
    data['l_time'] = Util.Date('%Y-%m-%d %H:%M:%S')
    token = Safety.Encode(data)
    # 缓存
    redis = Redis()
    key = Env.admin_token_prefix+'_token_'+str(data['uid'])
    redis.Set(key, '1')
    redis.Expire(key, Env.admin_token_time)
    redis.Close()
    return token
    
  # 获取
  def token(token: str):
    tData = Safety.Decode(token)
    if tData :
      redis = Redis()
      tData['time'] = redis.Ttl(Env.admin_token_prefix+'_token_'+tData['uid'])
      redis.Close()
    return tData

File: gin/service/admin_token.go
package service

import (
	"webmis/config"
	"webmis/library"
	"webmis/model"
	"webmis/util"

	"github.com/dgrijalva/jwt-go"
)

/* AdminToken */
type AdminToken struct {
	Base
}

/* 验证 */
func (s AdminToken) Verify(token string, urlPerm string) string {
	var redis *library.Redis
	// Token
	if token == "" {
		return "Token不能为空!"
	}
	tData, _ := (&library.Safety{}).Decode(token)
	if tData == nil {
		return "Token验证失败!"
	}
	// 是否过期
	env := config.Env()
	uid := (&util.Type{}).Strval(tData["uid"])
	key := env.AdminTokenPrefix + "_token_" + uid
	redis = (&library.Redis{}).New("")
	access_token := redis.Get(key)
	time := redis.TTL(key)
	redis.Close()
	if (&util.Hash{}).Md5(token) != access_token {
		return "强制退出!"
	}
	if time < 1 {
		return "Token已过期!"
	}
	// 续期
	if env.AdminTokenAuto {
		redis = (&library.Redis{}).New("")
		redis.Expire(key, env.AdminTokenTime)
		redis.Expire(env.AdminTokenPrefix+"_perm_"+uid, env.AdminTokenTime)
		redis.Close()
	}
	// URL权限
	if urlPerm == "" {
		return ""
	}
	arr := util.Explode("/", urlPerm)
	action := arr[len(arr)-1:][0]
	controller := util.Implode("/", arr[:len(arr)-1])
	// 菜单
	menu := (&model.SysMenu{}).New()
	menu.Columns("id", "action")
	menu.Where("controller=?", controller)
	menuData := menu.FindFirst()
	if len(menuData) == 0 {
		return "菜单验证无效!"
	}
	// 验证-菜单
	id := (&util.Type{}).Strval(menuData["id"])
	permData := s.Perm(token)
	actionVal, ok := permData[id]
	if !ok {
		return "无权访问菜单!"
	}
	// 验证-动作
	permArr := []map[string]interface{}{}
	util.JsonDecode(menuData["action"], &permArr)
	var permVal int64
	for _, val := range permArr {
		if action == val["action"] {
			permVal = (&util.Type{}).Int64(val["perm"])
			break
		}
	}
	if actionVal&permVal == 0 {
		return "无权访问动作!"
	}
	return ""
}

/* 权限数组 */
func (AdminToken) Perm(token string) map[string]int64 {
	permAll := map[string]int64{}
	// Token
	tData, _ := (&library.Safety{}).Decode(token)
	if tData == nil {
		return permAll
	}
	// 权限
	env := config.Env()
	redis := (&library.Redis{}).New("")
	permStr := redis.Get(env.AdminTokenPrefix + "_perm_" + (&util.Type{}).Strval(tData["uid"]))
	redis.Close()
	// 拆分
	arr := []string{}
	if !util.Empty(permStr) {
		arr = util.Explode(" ", permStr)
	}
	for _, val := range arr {
		s := util.Explode(":", val)
		permAll[s[0]] = (&util.Type{}).Int64(s[1])
	}
	return permAll
}

/* 生成 */
func (AdminToken) Create(data map[string]interface{}) string {
	data["l_time"] = util.Date("2006-01-02 15:04:05")
	token, _ := (&library.Safety{}).Encode(data)
	// 缓存
	env := config.Env()
	redis := (&library.Redis{}).New("")
	key := env.AdminTokenPrefix + "_token_" + (&util.Type{}).Strval(data["uid"])
	redis.Set(key, (&util.Hash{}).Md5(token))
	redis.Expire(key, env.AdminTokenTime)
	redis.Close()
	return token
}

/* 解析 */
func (AdminToken) Token(token string) jwt.MapClaims {
	tData, _ := (&library.Safety{}).Decode(token)
	if tData != nil {
		env := config.Env()
		redis := (&library.Redis{}).New("")
		tData["time"] = redis.TTL(env.AdminTokenPrefix + "_token_" + (&util.Type{}).Strval(tData["uid"]))
		redis.Close()
	}
	return tData
}
